Buy/Sell naming refactor

RoboRambo opened the pull request:
Resolves #5 

renamed #calculate_buy_price to #calculate_in_given_out & #calculate_sell_price to #calculate_out_given_in
adjusted tests & docs accordingly

RoboRambo committed:
renamed #calculate_buy_price to #calculate_in_given_out & #calculate_sell_price to #calculate_out_given_in - adjusted tests & docs accordingly

enthusiastmartin commented on src/amm.rs:
```diff
@@ -67,46 +67,45 @@ pub fn calculate_spot_price(sell_reserve: Balance, buy_reserve: Balance, amount:
 /// Returns None in case of error
-pub fn calculate_buy_price(sell_reserve: Balance, buy_reserve: Balance, amount: Balance) -> Option<Balance> {
-    ensure!(amount <= buy_reserve);
+pub fn calculate_in_given_out(out_reserve: Balance, in_reserve: Balance, amount_out: Balance) -> Option<Balance> {
```
Why the order of parameters has changed ?

jak-pan replied:
I think I have wrote that in description. We should make a consensus to do the order of params same everywhere. I know that it was something like action(action_asset, acted_asset, action_amount):acted_amount but right here it was not the case. We might want to sync it everywhere since we are doing cleanup. I'm good with deciding to go in then out always but we should state it somewhere and use it.

enthusiastmartin replied:
Ah right. Yes , correct. I only noticed it now that it was not correct here originally. So all good.

enthusiastmartin commented on src/tests.rs:
```diff
@@ -33,16 +33,16 @@ fn sell_price_should_work() {
+fn in_given_out_should_work() {
     let cases = vec![
-        (1000, 2000, 500, Some(334), "Easy case"),
+        (2000, 1000, 500, Some(334), "Easy case"),
```
Due to the change of order of params, this has changed. But I wonder if there is good reason to do as it would required more effort than just renaming the func where it is currently used.

RoboRambo replied:
I implemented it as mentioned in the issue, there the parameters were
switched. I wanted to point this out but forgot.

I can order it the way you think it makes sense.

enthusiastmartin replied:
no.all good

RoboRambo commented:
I implemented it as mentioned in the issue, there the parameters were
switched. I wanted to point this out but forgot.

I can order it the way you think it makes sense.

On Tue, 23 Feb 2021, 14:39 Martin <notifications@github.com> wrote:

> *@enthusiastmartin* commented on this pull request.
> ------------------------------
>
> In src/amm.rs
> <https://github.com/galacticcouncil/HydraDX-math/pull/8#discussion_r581040100>
> :
>
> >  ///
>  /// - `sell_reserve` - reserve amount of selling asset
>  /// - `buy_reserve` - reserve amount of buying asset
>  /// - `amount` - buy amount
>  ///
>  /// Returns None in case of error
> -pub fn calculate_buy_price(sell_reserve: Balance, buy_reserve: Balance, amount: Balance) -> Option<Balance> {
> -    ensure!(amount <= buy_reserve);
> +pub fn calculate_in_given_out(out_reserve: Balance, in_reserve: Balance, amount_out: Balance) -> Option<Balance> {
>
> Why the order of parameters has changed ?
> ------------------------------
>
> In src/tests.rs
> <https://github.com/galacticcouncil/HydraDX-math/pull/8#discussion_r581041033>
> :
>
> >      let cases = vec![
> -        (1000, 2000, 500, Some(334), "Easy case"),
> +        (2000, 1000, 500, Some(334), "Easy case"),
>
> Due to the change of order of params, this has changed. But I wonder if
> there is good reason to do as it would required more effort than just
> renaming the func where it is currently used.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/galacticcouncil/HydraDX-math/pull/8#pullrequestreview-596338621>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADDDCITWXVYJJKOHAXFCTVLTAOVZLANCNFSM4YBNWSQQ>
> .
>

jak-pan commented:
Let's bump the package version. We should start versioning since more people are joining in. We'll need another companion for this in WASM and HydraDX (master and parachain if we want latest stuff)

RoboRambo commented:
Should i change the version in cargo.toml?
If so please suggest a value.

jak-pan commented:
If we're going with semver this is a breaking change so 1.0.0

RoboRambo committed:
changed version number to 1.0.0

RoboRambo committed:
Merge branch 'main' into change_sell_buy_to_in_out

enthusiastmartin commented:
> If we're going with semver this is a breaking change so 1.0.0

with the new version - we need to be careful when merging this to master. Because it will immediately break the repos which are using this crate because they use latest master ( that means they wont find 0.1.0 ).

So to make things easier to integrate and upgrade, we can either :

1. in the repos  (node,wasm) - use specific revision with 0.1.0 (tag would work too ) so all continues to work
2. create a branch in this repo - eg v1.0.0 - where we merge all the breaking changes, once all ready, we can update other repos and merge to master.

jak-pan commented:
I think that tags are the easiest right now.

enthusiastmartin commented:
> I think that tags are the easiest right now.

done with revision. I created a tag for the initial release but for some reason, cargo could not find it. So i used the revision associated with the tag.  Basically the same.

So we should be ok merging breaking changes to main/master.

enthusiastmartin commented on src/amm.rs:
```diff
@@ -56,46 +56,45 @@ pub fn calculate_spot_price(sell_reserve: Balance, buy_reserve: Balance, amount:
-/// Formula : BUY_RESERVE * AMOUNT / (SELL_RESERVE + AMOUNT )
+/// Formula : OUT_RESERVE * AMOUNT_IN / (IN_RESERVE + AMOUNT_IN)
 ///
 /// - `sell_reserve` - reserve amount of selling asset
```
also this needs an update. perhaps also a better docs format and wording

enthusiastmartin commented on src/amm.rs:
```diff
@@ -56,46 +56,45 @@ pub fn calculate_spot_price(sell_reserve: Balance, buy_reserve: Balance, amount:
+/// Formula : OUT_RESERVE * AMOUNT_IN / (IN_RESERVE + AMOUNT_IN)
 ///
 /// - `sell_reserve` - reserve amount of selling asset
 /// - `buy_reserve` - reserve amount of buying asset
```
same here etc

enthusiastmartin commented on src/amm.rs:
```diff
@@ -56,46 +56,45 @@ pub fn calculate_spot_price(sell_reserve: Balance, buy_reserve: Balance, amount:
+/// Formula : IN_RESERVE * AMOUNT_OUT / (OUT_RESERVE - AMOUNT_OUT)
 ///
 /// - `sell_reserve` - reserve amount of selling asset
 /// - `buy_reserve` - reserve amount of buying asset
```
same here.

enthusiastmartin commented on src/tests.rs:
```diff
@@ -24,7 +24,7 @@ fn sell_price_should_work() {
 
     for case in cases {
         assert_eq!(
-            crate::amm::calculate_sell_price(case.0, case.1, case.2),
```
perhaps also test names should be updated.

RoboRambo committed:
renamed test & adjusted docs

RoboRambo committed:
Merge remote-tracking branch 'origin/change_sell_buy_to_in_out' into change_sell_buy_to_in_out

jak-pan commented on src/amm.rs:
```diff
@@ -56,46 +56,45 @@ pub fn calculate_spot_price(sell_reserve: Balance, buy_reserve: Balance, amount:
 }
 
 /// Calculating selling price given reserve of selling asset and reserve of buying asset.
```
```suggestion
/// Calculate amount to be received from the pool given the amount to be sent to the pool and both reserves.
```

jak-pan commented on src/amm.rs:
```diff
@@ -56,46 +56,45 @@ pub fn calculate_spot_price(sell_reserve: Balance, buy_reserve: Balance, amount:
     }
 }
 
 /// Calculating buying price given reserve of selling asset and reserve of buying asset.
```
```suggestion
/// Calculate amount to be sent to the pool given the amount to be received from the pool and both reserves.
```

jak-pan reviewed (commented):
LGTM after https://github.com/galacticcouncil/HydraDX-node/pull/115 merges and two suggestions.

RoboRambo committed:
adjusted docs

jak-pan commented on src/amm.rs:
```diff
@@ -34,7 +34,7 @@ macro_rules! to_u128 {
     };
 }
 
-/// Calculating spot price given reserve of selling asset and reserve of buying asset.
```
Completely missed this function. Let's also do this one like every other. (sell = in, buy = out)

jak-pan replied:
time to sleep for me :D

RoboRambo committed:
adjusted docs & param names in signature of #calculate_spot_price

enthusiastmartin committed:
Merge branch 'main' into change_sell_buy_to_in_out

enthusiastmartin commented on src/amm.rs:
```diff
@@ -34,18 +34,18 @@ macro_rules! to_u128 {
+    ensure!(in_reserve != 0);
 
-    let (amount_hp, buy_reserve_hp, sell_reserve_hp) = to_u256!(amount, buy_reserve, sell_reserve);
+    let (amount_hp, buy_reserve_hp, sell_reserve_hp) = to_u256!(amount, out_reserve, in_reserve);
```
perhaps it would good to change also local variables with in/out.

enthusiastmartin reviewed (commented):
Just the local variables so there wont be any buy/sell at all. and ready to merge.

RoboRambo committed:
changed variable names & adjusted docs

enthusiastmartin reviewed (approved):
Looks good. I'll merge soon.
